feat: embedded playground (#68)
openfga/openfga: README.md

[![Go Reference](https://pkg.go.dev/badge/github.com/openfga/openfga.svg)](https://pkg.go.dev/github.com/openfga/openfga)
![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/openfga/openfga?sort=semver&color=green)
[![Container Image](https://img.shields.io/github/v/release/openfga/openfga?color=blueviolet&label=container&logo=docker "Container Image")](https://hub.docker.com/r/openfga/openfga/tags)

> ![Downloads](https://img.shields.io/github/downloads/openfga/openfga/total.svg?style=flat&color=lightgrey) <!-- render new line with double space here -->  

Reviewer: If we are removing the double space then can remove the comment as well please?

Author: sure!

![Downloads](https://img.shields.io/github/downloads/openfga/openfga/total.svg?style=flat&color=lightgrey)
[![License](https://img.shields.io/badge/License-Apache_2.0-blue.svg)](./LICENSE)
![Codecov](https://img.shields.io/codecov/c/github/openfga/openfga)
[![Go Report](https://goreportcard.com/badge/github.com/openfga/openfga)](https://goreportcard.com/report/github.com/openfga/openfga)
![Snyk Vulnerabilities for GitHub Repo](https://img.shields.io/snyk/vulnerabilities/github/openfga/openfga?color=orange)
[![Discord Server](https://img.shields.io/discord/759188666072825867?color=7289da&logo=discord "Discord Server")](https://discord.com/channels/759188666072825867/930524706854031421)
[![Twitter](https://img.shields.io/twitter/follow/openfga?color=%23179CF0&logo=twitter&style=flat-square "@openfga on Twitter")](https://twitter.com/openfga)

A high-performance and flexible authorization/permission engine built for developers and inspired by [Google Zanzibar](https://research.google/pubs/pub48190/).

OpenFGA is designed to make it easy for developers to model their application permissions and add and integrate fine-grained authorization into their applications.

> ## Table of Contents

Reviewer: I am wondering if we need a ToC. I am not sure how much value it adds, and it seems like another thing that will quickly get out of sync.

Actually, I just found that a ToC is automatically generated. https://github.blog/changelog/2021-04-13-table-of-contents-support-in-markdown-files/

Reviewer 2: Nice 👍 !

Reviewer 2: Dropped the TOC in favor of the autogenerated one.

Reviewer 2: Dropped the TOC in favor of the autogenerated one.

- [Getting Started](#getting-started)
  - [Setup and Installation](#setup-and-installation)
    - [Docker](#docker)
    - [Pre-compiled Binaries](#pre-compiled-binaries)
    - [Source](#building-from-source)
  - [Verifying the Installation](#verifying-the-installation)
  - [Next Steps](#next-steps)
- [Playground](#playground)
- [Production Readiness](#production-readiness)
- [Contributing](#contributing)

## Getting Started

The following section aims to help you get started quickly. Please look at our official [documentation](https://openfga.dev/) for in-depth information.

### Setup and Installation

> > ℹ️ The following sections setup an OpenFGA server using the default configuration values. These are for rapid development and not for a production environment. Data written to an OpenFGA instance with default configurations will *not* persist after the service is stopped.

Reviewer: A little more explanation here maybe?
```suggestion
> ℹ️ The following sections setup an OpenFGA server using the default configuration values. These are for rapid development and not for a production environment. Data written to an OpenFGA instance using the default configuration with the memory storage engine will *not* persist after the service is stopped.
```

This section assumes that you have cloned the repository.

To run OpenFGA with the Postgres datastore engine, simply run the following commands:

```bash
docker compose up -d postgres
make run-postgres
```

This should start a Postgres container, write the database schema, and start the OpenFGA server.

When you are done you can stop the Postgres container with:

> ````bash

Reviewer: ```suggestion
```bash
```

Reviewer: Also I think we are not based off main, because there should be a closing ``` after the `docker compose down`.

--data-raw '{
    "name": "openfga-demo"
}'

> ````

Reviewer: ```suggestion
```
```
Or does four ticks work? Is this intentional?

}
```

## Playground

OpenFGA includes a playground use to facilitate the creation and visualisation of authorization models and tuples within your running local instance.

Once OpenFGA is running, the playground can be accessed at [http://localhost:3000/playground].

> In the event that a port other than the default `:3000` is required, the `OPENFGA_PLAYGROUND_PORT` environment can be set when starting the instance.

Reviewer: ```suggestion
In the event that a port other than the default `:3000` is required, the `OPENFGA_PLAYGROUND_PORT` environment variable can be set when starting the instance.
```

> eg:

Reviewer: I think this is getting confusing having so many examples of running OpenFGA. I think most people know how to set env vars, so perhaps the example can be dropped? WDYT?

Author: I assume nothing!

Reviewer: lol. If you want to keep them can we move the `eg:` to the same line then?

> ... starting the instance. E.g.,

Author: Sure!

}
```

## Playground

OpenFGA includes a playground use to facilitate the creation and visualisation of authorization models and tuples within your running local instance.

Once OpenFGA is running, the playground can be accessed at [http://localhost:3000/playground].

In the event that a port other than the default `:3000` is required, the `OPENFGA_PLAYGROUND_PORT` environment variable can be set when starting the instance. E.g.,

**Pre-compiled Binaries**

```sh
OPENFGA_PLAYGROUND_PORT=3001 ./bin/openfga
```

**Source**

```sh
OPENFGA_PLAYGROUND_PORT=3001 ./openfga
```

> 

Reviewer: Should we add a docker section with:

```sh
docker run -p 8080:8080 -p 3000:3000 openfga/openfga run
```

Although, I'll mention it again, I think we should simplify our README and the many ways of running and building that we are documenting. I think we should make use of our Makefile and let that be the documentation. What does everyone think?

Reviewer 2: Personally, I don't love make target bloat such as `make run`, `make run-playground`, `make run-postgres`, `make run-otherA`, etc.. Config should be easy to change and manage, and if it isn't then we should be prioritizing that problem.

I think the Makefile's "North Star" should always be to assist with common, and tedious, developer setup and development tasks. It shouldn't just become a dumping ground for alternate make targets that just set a config value differently before invoking the binary. 

This is obviously my opinion, and I think it's worth talking about more generally, but I think we should strive to discourage make targets that are just aliases for running the server with a particular config value. I think it's ok to ask a developer to change config if they want to run the server with different config.

Reviewer: Actually, I agree with you. What I was thinking more of is that we know have quite a number of ways to run OpenFGA in the readme and, in my opinion, it would be ideal to have (the Docker way of course and then) one easy way such as `make run`, and that would build and execute. If a user needs more details they can take a look at the Makefile. So the Makefile documents the steps to build from source, and we can remove that from the README, where I think it is not necessary.

Once we have an easy to set config (as in your PR) then it makes sense to to trim what we have now in the Makefile. As it is currently, since we don't read from a `.env` file, and finding out what env variables we can set, I think having a bunch of Make targets is not actually that bad a compromise.

Reviewer 2: > As it is currently, since we don't read from a .env file, and finding out what env variables we can set, I think having a bunch of Make targets is not actually that bad a compromise.

I agree. I think once we get these config changes we can start moving away from this pattern like you said as well.

> Actually, I agree with you. What I was thinking more of is that we know have quite a number of ways to run OpenFGA in the readme and, in my opinion, it would be ideal to have (the Docker way of course and then) one easy way such as make run, and that would build and execute.

I get what you're saying, and I like the intent behind it a lot. I think the challenges comes from the fact that we have different installation/setup methods, and having a docker target and a `make run` target leaves out the folks that may want to run the pre-compiled binary. I think for the source code installation instructions we could simply focus on the `make run` approach. But we do have

* Docker
* Pre-compiled Binaries
* Building from Source

I think we could probably drop the `docker-compose` instructions and just prioritize the docker instructions, but drop a hint that you can use the docker-compose file included in the repository if you like to use docker-compose 🤷 ?
